File: src/search/flood_fill.py
from __future__ import annotations

from collections import deque
from math import ceil
from typing import List

from src.core.types import Position


def safe_reachable_cells(search, pacman: Position, ghost: Position, max_depth: int = 16, ghost_speed: int = 2) -> List[Position]:
    ghost_dist = search.bfs_distance_map(ghost)
    seen = {pacman}
    q = deque([(pacman, 0)])
    cells: List[Position] = []

    while q:
        p, t = q.popleft()
        gd = ghost_dist[p[0]][p[1]]
        ghost_catch_turn = 0 if gd < 2 else ceil((gd - 1) / max(1, ghost_speed))
        if t >= ghost_catch_turn:
            continue
        cells.append(p)
        if t >= max_depth:
            continue
        for n in search.neighbors(p):
            if n not in seen:
                seen.add(n)
                q.append((n, t + 1))
    return cells


def safe_reachable_area(search, pacman: Position, ghost: Position, max_depth: int = 16, ghost_speed: int = 2) -> int:
    return len(safe_reachable_cells(search, pacman, ghost, max_depth, ghost_speed))


def reachable_area(search, start: Position, max_depth: int = 18) -> int:
    seen = {start}
    q = deque([(start, 0)])
    count = 0
    while q:
        p, d = q.popleft()
        count += 1
        if d >= max_depth:
            continue
        for n in search.neighbors(p):
            if n not in seen:
                seen.add(n)
                q.append((n, d + 1))
    return count

File: ts-backend/src/search/floodFill.ts
import type { Position } from "../types.js";
import { INF } from "../types.js";
import { isOpen, neighbors } from "../movement.js";
import type { SearchBoard } from "./bfs.js";

export function reachableArea(board: SearchBoard, start: Position, maxDepth = 18): number {
  if (!isOpen(board.grid, start)) return 0;
  const queue: Array<[Position, number]> = [[start, 0]];
  const seen = new Set<string>([`${start[0]},${start[1]}`]);
  let head = 0;
  while (head < queue.length) {
    const [pos, depth] = queue[head++];
    if (depth >= maxDepth) continue;
    for (const [next] of neighbors(board.grid, pos)) {
      const k = `${next[0]},${next[1]}`;
      if (!seen.has(k)) {
        seen.add(k);
        queue.push([next, depth + 1]);
      }
    }
  }
  return seen.size;
}

export function safeArea(board: SearchBoard, ghost: Position, pacman: Position, maxDepth = 18, pacmanSpeed = 2): number {
  if (!isOpen(board.grid, ghost)) return 0;
  const pacDist = board.distanceMap(pacman);
  const queue: Array<[Position, number]> = [[ghost, 0]];
  const seen = new Set<string>([`${ghost[0]},${ghost[1]}`]);
  let safe = 0;
  let head = 0;
  while (head < queue.length) {
    const [pos, depth] = queue[head++];
    const d = pacDist[pos[0]][pos[1]];
    const pacTurns = d >= INF ? INF : Math.ceil(d / Math.max(1, pacmanSpeed));
    if (pacTurns <= depth + 1) continue;
    safe += 1;
    if (depth >= maxDepth) continue;
    for (const [next] of neighbors(board.grid, pos)) {
      const k = `${next[0]},${next[1]}`;
      if (!seen.has(k)) {
        seen.add(k);
        queue.push([next, depth + 1]);
      }
    }
  }
  return safe;
}
